servant-client/test/Servant/ClientSpec.hs

```diff
@@ -148,6 +154,46 @@ failServer = serve failApi EmptyConfig (
   :<|> (\_request respond -> respond $ responseLBS ok200 [("content-type", "fooooo")] "")
  )
 
+--  auth stuff
+type AuthAPI =
+       BasicAuth "foo-realm" :> "private" :> "basic" :> Get '[JSON] Person
+  :<|> AuthProtect "auth-tag" :> "private" :> "auth" :> Get '[JSON] Person
+
+authAPI :: Proxy AuthAPI
+authAPI = Proxy
+
+type instance AuthReturnType (BasicAuth "foo-realm")  = ()
+type instance AuthReturnType (AuthProtect "auth-tag") = ()
+type instance AuthClientData (AuthProtect "auth-tag") = ()
+
+basicAuthHandler :: BasicAuthCheck ()
+basicAuthHandler =
+  let check (BasicAuthData username password) =
+        if username == "servant" && password == "server"
+        then return (Authorized ())
+        else return Unauthorized
+  in BasicAuthCheck check
+
+authHandler :: AuthHandler Request ()
+authHandler =
+  let handler req = case lookup "AuthHeader" (requestHeaders req) of
+        Nothing -> throwE (err401 { errBody = "Missing auth header"
+                                  , errReasonPhrase = "denied!"
```

Comment:
I am partly doing an end-to-end test here, to ensure that the way the serve modify's the response is verified by the client. I can change this to only test the body.